streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/StickyTaskAssignor.java

```diff
@@ -0,0 +1,145 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.processor.internals.assignment;
+
+import org.apache.kafka.streams.processor.TaskId;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+public class StickyTaskAssignor<ID> implements TaskAssignor<ID, TaskId> {
+
+    private final Map<ID, ClientState<TaskId>> clients;
+    private final Set<TaskId> taskIds;
+    private final Map<TaskId, ID> previousActiveTaskAssignment = new HashMap<>();
+    private final Map<TaskId, Set<ID>> previousStandbyTaskAssignment = new HashMap<>();
+
+    public StickyTaskAssignor(final Map<ID, ClientState<TaskId>> clients, final Set<TaskId> taskIds) {
+        this.clients = clients;
+        this.taskIds = taskIds;
+        mapPreviousTaskAssignment(clients);
+    }
+
+    @Override
+    public void assign(final int numStandbyReplicas) {
+        assignActive();
+        assignStandby(numStandbyReplicas);
+    }
+
+    // Visible for testing
+    void assignStandby(final int numStandbyReplicas) {
+        for (int i = 0; i < numStandbyReplicas; i++) {
+            for (final TaskId taskId : taskIds) {
+                final Set<ID> ids = findClientsWithoutAssignedTask(taskId);
+                if (ids.isEmpty()) {
+                    continue;
+                }
+                final ClientState<TaskId> client = findClient(taskId, ids);
```

Comment:
In the old approach, we use `taskPairs` trying to distribute the standby tasks to better tolerate failures; for example if we have four clients, four tasks, and replica.num = 1, we would rather have

```
client1: active: t1, standby: t2
client2: active: t2, standby: t3
client3: active: t3, standby: t4
client4: active: t4, standby: t1
```

Instead of

```
client1: active: t1, standby: t2
client2: active: t2, standby: t1
client3: active: t3, standby: t4
client4: active: t4, standby: t3
```

since otherwise if we lose client1 and client2, we lose t2 only in the first case but will lose both t1 and t2 in the second case.